### PR TITLE
feat: automatic record linkage for CVEs

### PR DESCRIPTION
**Pull request by RaitoBezarius**
This is a v1 of the record linkage system.

This PR contains (or will):

- some design documents for decisions
- an implementation of an automatic record linkage

Fixes https://github.com/Nix-Security-WG/nix-security-tracker/issues/221.

No UI to validate proposals yet.

Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

### CODE REVIEW

File: docs/design/02_cpe-based_linkage.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+### Models
+
+When a matching is deemed as fit, it is inserted as a `CVEDerivationClusterProposals`.
+
+### Validation of matchings
+
+The security team can then review the new proposed matching of the day based on the new CVEs and can review also unmatched CVEs to triage them successfully as well.
+
+When a matching is validated, `CVEDerivationClusterProposal` moves its status to `ACCEPTED`, this generates in turn a Nixpkgs security issue and the rest of the workflow is the one for a manual linkage.
```
Comment (fricklerhandwerk) [inline docs/design/02_cpe-based_linkage.md:46]: With @erictapen we had discussed the alternative that we'd create a Nixpkgs  security issue with "suggestion" state directly, and then go from there. WDYT?

What's still missing for me in this document is an explicit statement for what it means when a suggestion gets dismissed: Will the CVE never be revisited again (except manually)?

Comment (fricklerhandwerk) [reply docs/design/02_cpe-based_linkage.md:46]: Discussed with @RaitoBezarius: Nixpkgs security issues are mutable, but we want to keep the matchings and whether they were rejected or accepted separately to be able to [train ML models](https://github.com/Nix-Security-WG/nix-security-tracker/issues/216) to improve automatic suggestions.

File: docs/design/01_linkage.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+In [34]: 299806/407168
+Out[34]: 0.7363201430367808
+```
+
+73 % of the affected products in the database, at the time of writing, have no CPEs.
+
+## Conclusion
+
+71 % of the CVE have a CPE string.
```
Comment (fricklerhandwerk) [inline docs/design/01_linkage.md:79]: This is missing the inference step, not clear how this number comes up

File: docs/design/01_linkage.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+We would like to find derivations which correspond to product `Z` and vendor `Y` in nixpkgs, details on the matching are defined in [the CPE-based linkage document](./02_cpe-based_linkage.md).
+
+[1]: https://nvlpubs.nist.gov/nistpubs/Legacy/IR/nistir7695.pdf
+
+### Without CPEs
+
+To make it simple, we should do the work to attach CPEs to CVEs themselves when they are missing them.
+
+We reduce this problem to predict or infer a valid CPE albeit wrong from a text description.
```
Comment (fricklerhandwerk) [inline docs/design/01_linkage.md:37]: Not clear to me why we wouldn't jump to derivation matching directly. CPEs seem not very high fidelity information.

Comment (fricklerhandwerk) [reply docs/design/01_linkage.md:37]: Discussed with @RaitoBezarius: We're not losing anything by having an intermediate (generated or curated) CPE layer, but we gain the opportunity to verify and improve the matching quality.

File: docs/design/01_linkage.md (hunk 2 above)
Comment (RaitoBezarius) [reply docs/design/01_linkage.md]: Explained.

File: src/website/shared/listeners/automatic_linkage.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        )
+        for drv in drvs
+    ]
+
+    # We create all the set in one shot.
+    CVEDerivationClusterProposal.derivations.through.objects.bulk_create(drvs_throughs)
+
+    if drvs_throughs:
+        logger.info(
+            "Proposal for '%s' made: %d candidates.", container.cve, len(drvs_throughs)
```
Comment (fricklerhandwerk) [inline src/website/shared/listeners/automatic_linkage.py:65]:
```suggestion
            "Matching suggestion for '%s': %d derivations found.", container.cve, len(drvs_throughs)
```

File: src/website/shared/management/commands/propose_cve_links.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        parser.add_argument(
+            "delta_range_in_days",
+            help="Since when should we redo all the links in days? e.g. `5` for 5 days ago.",
+        )
+
+    def handle(self, *args: Any, **kwargs: Any) -> None:
+        _delta = kwargs["delta_range_in_days"]
+
+        try:
+            delta = datetime.timedelta(days=int(_delta))
```
Comment (fricklerhandwerk) [inline src/website/shared/management/commands/propose_cve_links.py:27]: We should be able to force it to an `int` at the argparse stage

File: src/website/shared/management/commands/propose_cve_links.py [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+            since_date = datetime.datetime.now() - delta
+        except ValueError:
+            raise CommandError(f"Not a valid delta format: {_delta}")
```
Comment (fricklerhandwerk) [inline src/website/shared/management/commands/propose_cve_links.py:30]: How can this go wrong except for the cast?

File: src/website/shared/management/commands/propose_cve_links.py (hunk 5 above)
Comment (fricklerhandwerk) [reply src/website/shared/management/commands/propose_cve_links.py:27]: Maybe even a positive int?

**Review by fricklerhandwerk — APPROVED** [review]
Nice, just a few nits.

File: src/website/shared/management/commands/propose_cve_links.py (hunk 5 above)
Comment (RaitoBezarius) [reply src/website/shared/management/commands/propose_cve_links.py:27]: That's right.

File: src/website/shared/management/commands/propose_cve_links.py (hunk 6 above)
Comment (RaitoBezarius) [reply src/website/shared/management/commands/propose_cve_links.py:30]: What happens if you put 365 * 1_000_000 years? :-)

File: src/website/shared/management/commands/propose_cve_links.py (hunk 5 above)
Comment (RaitoBezarius) [reply src/website/shared/management/commands/propose_cve_links.py:37]: Done.

File: src/website/shared/management/commands/propose_cve_links.py (hunk 6 above)
Comment (RaitoBezarius) [reply src/website/shared/management/commands/propose_cve_links.py:40]: I even fixed that, but I think we can keep the defensive programming :).

**Pull request merged**